indi-svbony/svbony_base.cpp

```diff
@@ -0,0 +1,1341 @@
+    {
+        LOGF_INFO("Invalid height request %d", h);
+        return false;
+    }
+
+    // ZWO rules are this: width%8 = 0, height%2 = 0
+    // if this condition is not met, we set it internally to slightly smaller values
+
+    if (warn_roi_width && subW % 8 > 0)
+    {
+        LOGF_INFO("Incompatible frame width %dpx. Reducing by %dpx.", subW, subW % 8);
+        warn_roi_width = false;
+    }
+    if (warn_roi_height && subH % 2 > 0)
+    {
+        LOGF_INFO("Incompatible frame height %dpx. Reducing by %dpx.", subH, subH % 2);
+        warn_roi_height = false;
+    }
+
+    subW -= subW % 8;
+    subH -= subH % 2;
+
+    LOGF_DEBUG("Frame ROI x:%d y:%d w:%d h:%d", subX, subY, subW, subH);
+
+    SVB_ERROR_CODE ret;
+
+    ret = SVBSetROIFormat(mCameraInfo.CameraID, subX, subY, subW, subH, binX);
+    if (ret != SVB_SUCCESS)
+    {
+        LOGF_ERROR("Failed to set ROI (%s).", Helpers::toString(ret));
+        return false;
+    }
+
+    mCurrentVideoFormat = getImageType();
+    switch (mCurrentVideoFormat)
+    {
+        case SVB_IMG_RAW16:
+            PrimaryCCD.setBPP(16);
+            break;
+
+        default:
+            PrimaryCCD.setBPP(8);
+            break;
+    }
+
+    SVBSetOutputImageType(mCameraInfo.CameraID, mCurrentVideoFormat);
+
+    // Set UNBINNED coords
+    PrimaryCCD.setFrame(subX * binX, subY * binY, subW * binX, subH * binY);
+
+    // Total bytes required for image buffer
+    auto nbuf = (subW * subH * static_cast<uint32_t>(PrimaryCCD.getBPP()) / 8) * ((getImageType() == SVB_IMG_RGB24) ? 3 : 1);
+
+    LOGF_DEBUG("Setting frame buffer size to %d bytes.", nbuf);
+    PrimaryCCD.setFrameBufferSize(nbuf);
+
+    // Always set BINNED size
+    Streamer->setSize(subW, subH);
+
+    return true;
+}
+
+bool SVBONYBase::UpdateCCDBin(int binx, int biny)
+{
+    INDI_UNUSED(biny);
+
+    PrimaryCCD.setBin(binx, binx);
+
+    return UpdateCCDFrame(PrimaryCCD.getSubX(), PrimaryCCD.getSubY(), PrimaryCCD.getSubW(), PrimaryCCD.getSubH());
+}
+
+/* Downloads the image from the CCD.
+ N.B. No processing is done on the image */
+int SVBONYBase::grabImage(float duration)
+{
+    SVB_ERROR_CODE ret = SVB_SUCCESS;
+
+    SVB_IMG_TYPE type = getImageType();
+
+    std::unique_lock<std::mutex> guard(ccdBufferLock);
+    uint8_t *image = PrimaryCCD.getFrameBuffer();
+    uint8_t *buffer = image;
+
+    uint16_t subW = PrimaryCCD.getSubW() / PrimaryCCD.getBinX();
+    uint16_t subH = PrimaryCCD.getSubH() / PrimaryCCD.getBinY();
+    int nChannels = (type == SVB_IMG_RGB24) ? 3 : 1;
+    size_t nTotalBytes = subW * subH * nChannels * (PrimaryCCD.getBPP() / 8);
+
+    if (type == SVB_IMG_RGB24)
+    {
+        buffer = static_cast<uint8_t *>(malloc(nTotalBytes));
+        if (buffer == nullptr)
+        {
+            LOGF_ERROR("%s: %d malloc failed (RGB 24).", getDeviceName());
+            return -1;
+        }
+    }
+
+    ret = SVBGetVideoData(mCameraInfo.CameraID, buffer, nTotalBytes, 1000);
```

Comment:
SVBGetVideoData must be retried when ret is SVB_ERROR_TIMEOUT.
It is necessary to be able to accept Abort processing at that time. Otherwise, it will not be able to be aborted while SVB_ERROR_TIMEOUT continues.